src/Sarif/Visitors/SarifCurrentToVersionOneVisitor.cs

```diff
@@ -1032,6 +1014,81 @@ private static string CreateFileDictionaryKey(FileData v2File, IList<FileData> v
             return logicalLocationsVersionOne;
         }
 
+        // In SARIF v1, run.resources.rules was a dictionary, whereas in v2 it is an array.
+        // Normally, the lookup key into the v1 rules dictionary is the rule id. But some
+        // tools allow multiple rules to have the same id. In that case we must synthesize
+        // a unique key for each rule with that id. We choose "<ruleId>-<n>", where <n> is
+        // 1 for the second occurrence, 2 for the third, and so on.
+        private static IDictionary<int, string> CreateV2RuleIndexToV1KeyMapping(IList<Rule> rules)
+        {
+            var v2RuleIndexToV1KeyMap = new Dictionary<int, string>();
+
+            if (rules != null)
+            {
+                // Keep track of how many distinct rules have each id.
+                var ruleIdToCountMap = new Dictionary<string, int>();
+
+                for (int i = 0; i < rules.Count; ++i)
+                {
+                    Rule rule = rules[i];
+                    if (rule.Id != null)
+                    {
+                        ruleIdToCountMap[rule.Id] = ruleIdToCountMap.ContainsKey(rule.Id)
+                            ? ruleIdToCountMap[rule.Id] + 1
+                            : 1;
+
+                        v2RuleIndexToV1KeyMap[i] = ruleIdToCountMap[rule.Id] == 1
+                            ? null
```

Comment:
why store null in some cases? instead of just persisting the rule id? then look-up just works exclusively against this look-up. or perhaps you use the null case to indicate something special later?

Reply:
BUT! I found what I think is a clearer way to express this logic. Now, I _do_ set the map entry to `ruleId` rather than `null`, and I _don't_ need the `null` check (and its comment) in `ConvertRulesArrayToDictionary`. Instead, I do the `null` check right at the point where we care about its effect: at the point where I assign `result.RuleKey`.

See the last commit.

---
In reply to: [249616588](https://github.com/Microsoft/sarif-sdk/pull/1228#discussion_r249616588) [](ancestors = 249616588,249615764,249573985)